rx_signal_processing/decimate.cu

```diff
@@ -1,221 +1,391 @@
-    filter_products[tap_offset] = cuCaddf(filter_products[tap_offset],filter_products[tap_offset+1]);
-
-    __syncthreads();
-    auto calculated_output_sample = parallel_reduce(filter_products, tap_offset); // REVIEW #0 pass new variable for offset in filter products so you are not passing only even values
-                                                                  // REPLY this can stay the same, but I missed a line of code that needs to happen first
-
-    //again, wrong for now.
-    if (threadIdx.x == 0) { // REVIEW #1 Explain how you're setting up the array of decimated samples
-        channel_offset = channel_num * samples_per_channel/dm_rate; // REVIEW #13 gridDimx is already samples_per_channel/dm_rate, use it instead
-        auto total_channels = blockDim.y; // REVIEW #0 This should be gridDim.y if you intend to use 'total_channels' as antennas (should use 'antennas')
-        auto freq_offset = threadIdx.y * total_channels; // REVIEW #0 still need to multiply by gridDim.x here to get index into proper location
-        auto total_offset = freq_offset + channel_offset + dec_sample_num;
-        decimated_samples[total_offset] = calculated_output_sample;
-    }
-}
-
-static dim3 create_grid(uint32_t num_samples, uint32_t dm_rate, uint32_t num_channels) // REVIEW #26 no more channels
-{
-    auto num_blocks_x = num_samples/dm_rate;
-    auto num_blocks_y = num_channels;
-    auto num_blocks_z = 1;
-    std::cout << "    Grid size: " << num_blocks_x << " x " << num_blocks_y << " x "
-        << num_blocks_z << std::endl;
-    dim3 dimGrid(num_blocks_x,num_blocks_y,num_blocks_z);
-
-    return dimGrid;
-}
-
-static dim3 create_block(uint32_t num_taps, uint32_t num_freqs)
-{
-    auto num_threads_x = num_taps;
-    auto num_threads_y = num_freqs;
-    auto num_threads_z = 1;
-    std::cout << "    Block size: " << num_threads_x << " x " << num_threads_y << " x "
-        << num_threads_z << std::endl;
-    dim3 dimBlock(num_threads_x,num_threads_y,num_threads_z);
-
-    return dimBlock;
+  // Since number of filter taps is calculated at runtime and we do not want to hardcode
+  // values, the shared memory can be dynamically initialized at invocation of the kernel.
+  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared
+
+  extern __shared__ cuComplex filter_products[];
+
+  auto antenna_num = blockIdx.y;
+  auto antenna_offset = antenna_num * samples_per_antenna;
+
+  auto dec_sample_num = blockIdx.x;
+  auto dec_sample_offset = dec_sample_num * dm_rate;
+
+  auto tap_offset = threadIdx.y * blockDim.x + 2 * threadIdx.x;
+
+  cuComplex sample_1;
+  cuComplex sample_2;
+  // If an offset should extend past the length of samples per antenna
+  // then zeroes are used as to not segfault or run into the next buffer.
+  // output samples using these will be discarded as to not introduce edge effects
```

Comment:
Reading this, we don't know where the removal of samples happens, would be good to know how the discarding works/where it happens